### ISSUE TITLE
Windows Taskbar grouping

### ISSUE THREAD

**Issue by fbordignon** [Slicer/Slicer#5616]
When I pin the Slicer launcher to the windows taskbar, it works for launching slicer but the actual slicerapp-real.exe icon is exhibited on a separate button. i.e. I end up with the shortcut for the launcher pinned and with the actual running slicer in another button. From what I gather the problem is that Slicer.exe and SlicerApp-real.exe are separate processes so windows treat them as separate programs from the taskbar perspective.

It would be great if the icons could be grouped to save taskbar space.

I've tried to add
```
#include<shobjidl_core.h>
SetCurrentProcessExplicitAppUserModelID(L"NAMIC.3DSlicer.4.11");
```
to the main functions of qSlicerApplicationMainWrapper.cxx and to Main.cpp of CTKAppLauncherLib with no success. They remain as separate icons at the taskbar. I am not sure of the exact position where the call need to be inserted.

Followed the instructions on https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-setcurrentprocessexplicitappusermodelid

**Comment by lassoan** [comment]
It would be a nice touch to group the launcher and the real executable into the same taskbar button.

I would recommend to use the Qt implementation (https://doc.qt.io/qt-5/qtwin.html#setCurrentProcessExplicitAppUserModelID) and make it the very first call. It is too late if you set it when a window is already displayed.

The launcher will have to read the model ID from the launcher settings file, because the exact same executable is used for all applications (building the launcher requires commercial Qt license due to building Qt statically).

If you pin a shortcut (and not a running application) then Windows cannot possibly know what the AppUserModelID would be of the running process. Probably you should set the AppUserModelID in the shortcut (see https://stackoverflow.com/questions/61806380/how-to-set-appusermodelid-of-application-pinned-to-taskbar). Since the shortcut is created during installation by NSIS, you need to find out how to specify the model ID for a shortcut using NSIS. Some potentially relevant pages: https://nsis.sourceforge.io/WinShell_plug-in, https://github.com/safing/nsis-shortcut-properties.

You may find this page useful: https://jcutrer.com/windows/find-aumid

**Comment by fbordignon** [comment]
Hi guys, I don't know if I need to work on this any further or if it is a matter of testing.
Let me know if you need anything more from my side.

Thanks.

**Comment by lassoan** [comment]
We'll try to get this reviewed and merged before the new stable release (within a few weeks).

**Comment by lassoan** [comment]
@fbordignon provided a partial solution in https://github.com/Slicer/Slicer/pull/5619, but there was a huge issue with that approach that it relied on a single global shortcut for all Slicer install trees of the same version.

An alternative solution, which fulfills the same requirements but does not need a global shortcut is the following:

1. Update the [CTK application launcher](https://github.com/commontk/AppLauncher) to set an AppUserModelID for the application and the console window.

```
  QString aumid = QString::fromLatin1("NA-MIC.Slicer.4.13.0-2021-12-01.X");
  std::wstring aumidw = aumid.toStdWString();
  SetCurrentProcessExplicitAppUserModelID(aumidw.c_str());

  HWND hwnd = GetConsoleWindow();
  IPropertyStore* pps;
  HRESULT hr = SHGetPropertyStoreForWindow(hwnd, IID_PPV_ARGS(&pps));
  if (SUCCEEDED(hr))
  {
    PROPVARIANT propvar;
    hr = InitPropVariantFromString(L"NA-MIC.Slicer.4.13.0-2021-12-01.X", &propvar);
    hr = pps->SetValue(PKEY_AppUserModel_ID, propvar);
    pps->Release();
  }
```

The AppUserModelID must be generated in a way that it provides a unique value for each install tree, even if the application version is exactly the same. For example, the AppUserModelID can be organization.applicationName.hash.
- organization would be used just to give an idea what created it
- applicationName can be used to create separate groups for each application specified in the launcher ini file (e.g., separate one for Slicer, designer, cmd, ...)
- hash should be generated from the absolute path of the install tree using MD5, which would allow separate pinned shortcuts for each install tree (even if they use the same application); there is a maximum length limitation of the AppUserModelID, so the hash may need to be trimmed, most likely it will be still unique enough.

2. Pass the AppUserModelID and relaunch command to the launched process. This is probably the simplest via environment variables.

3. In the launched process (SlicerApp-real.exe) set the AppUserModelID using SetCurrentProcessExplicitAppUserModelID; and set the relaunch command (launcher executable; with some additional arguments if not the main application is launched). Relaunch command setting is explained [here](https://stackoverflow.com/questions/57676759/how-to-allow-launcher-application-to-be-pinned-to-the-taskbar).

**Comment by fbordignon** [comment]
Thanks for the deep look into this issue. Indeed it is a bit more complicated.
I can confirm that pinning the running app generates a shortcut with the AUMID set, but it links to the SlicerApp-real.exe, which should be fixed by setting the relaunch command as described.
The unique AUMID is a good catch/idea as well, I am not a fan of passing things as environment variables. Is it too complicated to make it an argument when starting the launched process?

However, there is one or two cases where I could not find the solution: pin the Slicer.exe directly with right click on windows explorer:
![image](https://user-images.githubusercontent.com/4790289/144431835-976de61b-75c4-40df-8a64-81104c0f2dde.png)
That case is tricky because the user is pinning the launcher which will not know the AUMID until it is up and running. If the user pins that way, Slicer launches correctly but it does not group the windows. Perhaps we could add Slicer.exe to the [exclusion list ](https://docs.microsoft.com/en-us/windows/win32/shell/appids?redirectedfrom=MSDN#exclusion-lists-for-taskbar-pinning-and-recentfrequent-lists) so the user will be forced to pin only when the app is running. Adding Slicer.exe as a host app also has the effect of not allowing pinning.

Also, the user could pin the installed shortcut:
![image](https://user-images.githubusercontent.com/4790289/144438602-1a7d7c33-7f25-4ea7-8d72-808d362f4907.png)
which will be without AUMID (or with a global one). There is no way that I know of that the shortcut could compute the AUMID dynamically. My suggestion for the shortcut pinning is to use a nsh script to compute the AUMID at install time the same way the app computes it (I don't know if it is possible). if the user moves the installation folder the installed shortcut fails anyways.

**Comment by lassoan** [comment]
> The unique AUMID is a good catch/idea as well, I am not a fan of passing things as environment variables. Is it too complicated to make it an argument when starting the launched process?

I'm not a fan in general either, but this is a special case, because the launcher already sets environment variables extensively and has features such as storing the entire previous environment and able to recover the previous environment, so we don't pollute the environment by passing parameters like this (e.g., if the launched process uses a launcher then the previous value of the aumid environment variable will be preserved). The advantage of environment variables that we can launch processes that we own (and therefore can accept the additional two command-line arguments) and processes that we don't own (and therefore would not work properly if they got some unexpected arguments).

> That case is tricky because the user is pinning the launcher which will not know the AUMID until it is up and running. If the user pins that way, Slicer launches correctly but it does not group the windows.

Yes, indeed this might result in incorrect grouping. I have seen discussion about similar issues and one solution that someone implemented was that the launched application iterated through application shortcuts and updated the aumid in the shortcut that matched the launched application. It sounded quite invasive though, so I'm not sure if it is worth the trouble, but at least there is a potential solution if this becomes very important for someone.

**Comment by fbordignon** [comment]
I've just noticed this Note at [the docs](https://docs.microsoft.com/en-us/windows/win32/shell/appids?redirectedfrom=MSDN#where-to-assign-an-appusermodelid):

> If a shortcut exists to launch the application, an application should apply the AppUserModelID as a property of the shortcut instead of using the relaunch properties. In that case, the command line, icon, and text of the shortcut are used to supply the same information as the relaunch properties.

I did some testing and it seems that if there is a shortcut installed for the app, the taskbar uses the shortcut's AUMID and command to launch the app instead of the current executable. As it is described above. 
Hence, if we install the shortcut with the quasi-unique AUMID it should be fine, even if we set the AUMID at runtime only for SlicerApp-real.exe. In this case the taskbar will look for installed shortcuts with the same AUMID and use the installed shortcut. When pinning Slicer.exe, the taskbar will look for shortcuts pointing to the same Slicer.exe and use it. The only case that it will not work is when the user moves the installation folder and it messes with the shortcut.
We just need to know if NSIS can generate the same AUMID for the shortcut, considering the installation folder, as we would generate at runtime on SlicerApp-real.exe

**Comment by lassoan** [comment]
Yes, this is why pinning works nicely with the method in your pull request. However, it makes the application non-portable (non-relocatable) and does not work if you have the same Slicer version installed or copied to multiple locations on the same computer.
